src/main/java/workflow/context/AbstractCompositeContext.java

```diff
@@ -19,10 +19,12 @@
      * Instantiate a Context with a State
      * @param state ProcessState
      */
-    AbstractCompositeContext(ProcessState state, ContextInterface parentContext) {
+    AbstractCompositeContext(ProcessState state, ContextInterface parentContext, String key) {
         this.state = state;
         this.parent = parentContext;
-        this.childContexts = new HashMap<String, ContextInterface>();//TODO: should be a dictionary? will require streams work
+        this.key = key;
+        this.childContexts = new HashMap<String, ContextInterface>();
```

Comment:
Now I'm getting worried. I don't see how this won't become a mess.

Maybe there's a way we can relate parent/child keys in high level hash map? Maybe something like child_key = hash(parent_key + child_id)? 

I would like to see how this will be managed before I sign off on it.